File: python/2_add_two_numbers.py
"""
#2 - Add Two Numbers (Easy)

You are given two non-empty linked lists representing two non-negative integers. 

The digits are stored in reverse order and each of their nodes contain a single digit. 

Add the two numbers and return it as a linked list.

You may assume the two numbers do not contain any leading zero, except the number 0 itself.

Example:

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
Explanation: 342 + 465 = 807.

"""
#Definition for singly-linked list.
class ListNode(object):
	def __init__(self, x):
		self.val = x
		self.next = None
def addTwoNumbers(self, l1, l2):
	"""
	:type l1: ListNode
	:type l2: ListNode
	:rtype: ListNode
	"""
	"""
	Add two numbers bit by bit as usual, but be careful about carry and different situations.
	O(A+B) <- Loop through both linked list once
	"""
	# when one of the linked list is empty
	if l1 and not l2:
		return l1
	if l2 and not l1:
		return l2

	# save node in a list, 'up' saves the carry
	l3 = []
	up = 0
	# add up the num bit by bit
	while l1 and l2:
		# update the carry, remained is saved in l3
		add = l1.val + l2.val + up
		up = int(add >= 10)
		l3.append(ListNode(add % 10))
		if len(l3) > 1:
			# link them
			l3[-2].next = l3[-1]
		l1, l2 = l1.next, l2.next

	# if one of the linked list is exhausted, link remained to l3
	if l1 and not l2:
		l3[-1].next = l1
		l1.val += up
	if l2 and not l1:
		l3[-1].next = l2
		l2.val += up
	# if both exhausted, deal with the carry
	elif up:
		l3.append(ListNode(up))
		l3[-2].next = l3[-1]
	return l3[0]

File: C++/2_add_two_numbers.cpp
// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(nullptr) {}
};

#include <iostream>
#include <cstdlib>
using namespace std;

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        // handle special inputs
        if (!l1 && l2) 
            return l2;
        else if (!l2 && l1)
            return l1;
        else if (!l1 && !l2)
            return 0;
        
        // use a int for carry flag, a cycling node and a counter for ending
        int add = 0, count = 0;
        ListNode *endNode = new ListNode(0), *head = new ListNode(0), *p = head;
        endNode->next = endNode;

        // when both numbers are not exhausted
        while (count < 2) {
            int sum = l1->val + l2->val + add;
            add = sum / 10;    // update carry flag
            p->next = new ListNode(sum % 10);
            p = p->next;

            // if l1 or/and l2 ends, replace with the cycling node, and count
            if (!l1->next) {
                l1 = endNode;
                count += 1;
            }
            else
                l1 = l1->next;
            if (!l2->next) {
                l2 = endNode;
                count += 1;
            }
            else
                l2 = l2->next;
        }
        
        // handle the carry flag
        if (add)
            p->next = new ListNode(add);
        
        return head->next;
    }
};

// driver
int main(){
    ListNode *l1 = new ListNode(6), *l2 = new ListNode(5);
    Solution s;

    ListNode* l3 = s.addTwoNumbers(l1, l2);

    while(l3) {
        cout << l3->val << endl;
        l3 = l3->next;
    }

    system("pause");
    return 0;
}